satpy/readers/yaml_reader.py

```diff
@@ -1221,25 +1207,94 @@ def _load_dataset(dsid, ds_info, file_handlers, dim='y', pad_data=True):
+
     def _load_area_def(self, dsid, file_handlers, pad_data=True):
         """Load the area definition of *dsid* with padding."""
         if not pad_data:
             return _load_area_def(dsid, file_handlers)
-        return _load_area_def_with_padding(dsid, file_handlers)
+        return self._load_area_def_with_padding(dsid, file_handlers)
+
+    def _load_area_def_with_padding(self, dsid, file_handlers):
+        """Load the area definition of *dsid* with padding."""
+        # Pad missing segments between the first available and expected
+        area_defs = self._pad_later_segments_area(file_handlers, dsid)
 
+        # Add missing start segments
+        area_defs = self._pad_earlier_segments_area(file_handlers, dsid, area_defs)
 
-def _load_area_def_with_padding(dsid, file_handlers):
-    """Load the area definition of *dsid* with padding."""
-    # Pad missing segments between the first available and expected
-    area_defs = _pad_later_segments_area(file_handlers, dsid)
+        # Stack the area definitions
+        area_def = _stack_area_defs(area_defs)
 
-    # Add missing start segments
-    area_defs = _pad_earlier_segments_area(file_handlers, dsid, area_defs)
+        return area_def
 
-    # Stack the area definitions
-    area_def = _stack_area_defs(area_defs)
+    def _pad_later_segments_area(self, file_handlers, dsid):
+        """Pad area definitions for missing segments that are later in sequence than the first available."""
+        seg_size = None
+        expected_segments = file_handlers[0].filetype_info['expected_segments']
+        filetype = file_handlers[0].filetype_info['file_type']
+        available_segments = [int(fh.filename_info.get('segment', 1)) for
+                              fh in file_handlers]
+        area_defs = {}
 
-    return area_def
+        for segment in range(available_segments[0], expected_segments + 1):
+            try:
+                idx = available_segments.index(segment)
+                fh = file_handlers[idx]
+                area = fh.get_area_def(dsid)
+            except ValueError:
+                logger.debug("Padding to full disk with segment nr. %d", segment)
+
+                new_height_proj_coord, new_height_px = self._get_new_areadef_heights(area, seg_size,
+                                                                                     segment_n=segment,
+                                                                                     filetype=filetype)
+                new_ll_y = area.area_extent[1] + new_height_proj_coord
+                new_ur_y = area.area_extent[1]
+                fill_extent = (area.area_extent[0], new_ll_y,
+                               area.area_extent[2], new_ur_y)
+                area = AreaDefinition('fill', 'fill', 'fill', area.crs,
+                                      seg_size[1], new_height_px,
+                                      fill_extent)
+
+            area_defs[segment] = area
+            seg_size = area.shape
```

Comment:
Coudl this be a separate function? And further have the exception handling in yet another function.

Reply:
I refactored the exception handling, so that the method is much slimmer now. But I kept the try/except here, as refactoring further would introduce methods with many arguments, so not sure if it helps readability/makes much sense in the new code. What do you think?

Reply:
The whole `for` loop could be a separate method, say `return self._get_area_def_segments(filehandler, dsid, available_segments, expected_segments)` and move the `segment = None` initialisation also inside it.